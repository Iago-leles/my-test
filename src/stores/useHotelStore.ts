import { defineStore } from 'pinia'
import { ref, computed } from 'vue'

import type { Hotel } from '@/types'

import hotelData from '@/db/hotel.json'

import { HotelOrder } from '@/enums/hotels'

export const useHotelStore = defineStore('hotel', () => {
  const hotels = ref<Hotel[]>(hotelData.flatMap((data) => data.hotels) as Hotel[])
  const selectedHotel = ref<Hotel | null>(null)
  const isLoading = ref(false)
  const currentPage = ref(1)
  const itemsPerPage = 10
  const selectedPlace = ref<{ value: string; label: string } | null>(null)
  const orderBy = ref(HotelOrder.RECOMMENDED)
  const orderOptions = [HotelOrder.RECOMMENDED, HotelOrder.BEST_RATED]
  const filterName = ref('')

  const allHotels = computed(() => {
    return hotelData
  })

  function setFilterName(value: string) {
    filterName.value = value
  }

  const filteredHotels = computed(() => {
    let result = hotelData.flatMap((data) => data.hotels || []) as Hotel[]

    if (selectedPlace.value?.value) {
      result = allHotels.value
        .filter((places) => places.placeId === Number(selectedPlace.value?.value))
        .flatMap((places) => places.hotels) as Hotel[]
    }

    if (filterName.value.trim()) {
      const searchTerm = filterName.value.trim().toLowerCase()
      result = result.filter((hotel) => hotel.name.toLowerCase().includes(searchTerm))
    }

    const sortedOptions: Record<HotelOrder, Hotel[]> = {
      [HotelOrder.RECOMMENDED]: [...result].sort((a, b) => Number(a.price) - Number(b.price)),
      [HotelOrder.BEST_RATED]: [...result].sort((a, b) => Number(b.stars) - Number(a.stars)),
    }

    return sortedOptions[orderBy.value]
  })

  const displayedHotels = computed(() => {
    const endIndex = currentPage.value * itemsPerPage
    return filteredHotels.value.slice(0, endIndex)
  })

  function fetchNextPage(index: number, done: () => void) {
    setTimeout(() => {
      if (displayedHotels.value.length < filteredHotels.value.length) {
        currentPage.value++
        done()

        return
      }

      done()
    }, 500)
  }

  function setSelectedPlace(place: { value: string; label: string } | null) {
    selectedPlace.value = place || null
  }

  function setSelectedHotel(hotel: Hotel | null) {
    selectedHotel.value = hotel || null
  }

  return {
    hotels,
    isLoading,
    filterName,
    setFilterName,
    filteredHotels,
    displayedHotels,
    selectedPlace,
    orderOptions,
    orderBy,
    selectedHotel,
    fetchNextPage,
    setSelectedPlace,
    setSelectedHotel,
  }
})
