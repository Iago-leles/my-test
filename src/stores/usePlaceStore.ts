import { defineStore } from 'pinia'
import { ref, computed } from 'vue'

import type { Place } from 'src/types'

import placesData from 'src/db/place.json'

export const usePlaceStore = defineStore('place', () => {
  const places = ref<Place[]>(placesData)
  const selectedPlace = ref<{ label: string; value: string } | null>(null)
  const hasSearched = ref(false)

  const formattedPlaces = computed(() => {
    return places.value.map((place) => ({
      label: `${place.name}, ${place.state.name}`,
      value: String(place.placeId),
    }))
  })

  const selectedPlaceData = computed(() => {
    return formattedPlaces.value.find((place) => place.value === selectedPlace.value?.value)
  })

  function setSelectedPlace(place: { label: string; value: string } | null) {
    selectedPlace.value = place
  }

  function setHasSearched(value: boolean) {
    hasSearched.value = value
  }

  function resetSearch() {
    hasSearched.value = false
  }

  return {
    places,
    selectedPlace,
    hasSearched,
    formattedPlaces,
    selectedPlaceData,
    setSelectedPlace,
    setHasSearched,
    resetSearch,
  }
})
