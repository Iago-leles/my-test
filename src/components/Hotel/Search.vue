<template>
  <q-card class="hotel-search" flat bordered>
    <p class="text-h5 q-pa-sm q-ml-sm q-mb-none text-grey-8">{{ $t('hotels.search.title') }}</p>

    <q-separator color="grey-6" />

    <div class="q-pa-md full-width column wrap items-end">
      <div class="inputs">
        <app-select
          :label="$t('hotels.search.destination')"
          v-model="placeStore.selectedPlace"
          required
          outlined
          clearable
          dense
          display-value
          class="q-mb-md full-width"
          :options="placeStore.formattedPlaces"
        />

        <app-text-input
          :model="hotelStore.filterName"
          :label="$t('hotels.search.name')"
          required
          dense
          class="q-mb-md full-width"
          outlined
        />
      </div>

      <q-btn
        no-caps
        color="blue"
        class="q-px-xl q-py-xs"
        size="lg"
        rounded
        :label="buttonLabel"
        :loading="hotelStore.isLoading"
        @click="handleSearch"
      />
    </div>

    <q-inner-loading :showing="hotelStore.isLoading">
      <q-spinner-dots size="50px" color="primary" />
    </q-inner-loading>
  </q-card>
</template>

<script setup lang="ts">
import { computed } from 'vue'

import AppSelect from '@/components/ui/AppSelect.vue'
import AppTextInput from '@/components/ui/AppTextInput.vue'

import { usePlaceStore } from '@/stores/usePlaceStore'
import { useHotelStore } from '@/stores/useHotelStore'

import { useI18n } from 'vue-i18n'

const placeStore = usePlaceStore()
const hotelStore = useHotelStore()

const { t } = useI18n()

const buttonLabel = computed(() => {
  return placeStore.hasSearched ? t('hotels.search.button_alter') : t('hotels.search.button')
})

const handleSearch = () => {
  if (!placeStore.selectedPlace) {
    placeStore.setHasSearched(false)
    hotelStore.setSelectedPlace(null)

    return
  }

  placeStore.setHasSearched(true)
  hotelStore.setSelectedPlace(placeStore.selectedPlace)
}
</script>

<style scoped lang="scss">
.hotel-search {
  border-radius: 0.625rem;
  border: none;
}

.inputs {
  display: flex;
  gap: 1rem;
  width: 100%;
}
</style>
