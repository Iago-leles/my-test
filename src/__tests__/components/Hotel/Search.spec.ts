import { mount, flushPromises } from '@vue/test-utils'
import { expect, describe, it, vi } from 'vitest'
import { createTestingPinia } from '@pinia/testing'
import { useHotelStore } from '@/stores/useHotelStore'

import Search from '@/components/Hotel/Search.vue'

vi.mock('vue-i18n', () => ({
  useI18n: () => ({
    t: (key: string) => key,
  }),
}))

describe('Search', () => {
  it('should match snapshot', async () => {
    const pinia = createTestingPinia({
      createSpy: vi.fn,
      fakeApp: true,
    })

    const wrapper = mount(Search, {
      global: {
        plugins: [pinia],
      },
    })

    await flushPromises()

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should update the hotelStore filterName when input changes', async () => {
    const pinia = createTestingPinia({
      createSpy: vi.fn,
      stubActions: false,
    })

    const wrapper = mount(Search, {
      global: {
        plugins: [pinia],
      },
    })

    const input = wrapper.findComponent({ name: 'AppTextInput' })
    await input.setValue('Hotel Test')

    const hotelStore = useHotelStore(pinia)
    expect(hotelStore.filterName).toBe('Hotel Test')
  })

  it('should call handleSearch when button is clicked', async () => {
    const pinia = createTestingPinia({
      createSpy: vi.fn,
      stubActions: false,
    })

    const wrapper = mount(Search, {
      global: {
        plugins: [pinia],
      },
    })

    const button = wrapper.find('button')
    await button.trigger('click')

    const hotelStore = useHotelStore(pinia)
    expect(hotelStore.selectedPlace).toBeNull()
  })
})
