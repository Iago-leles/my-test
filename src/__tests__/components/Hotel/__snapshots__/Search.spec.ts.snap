// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Search > should match snapshot 1`] = `
"<div data-v-b2120978="" class="q-card q-card--bordered q-card--flat no-shadow hotel-search">
  <p data-v-b2120978="" class="text-h5 q-pa-sm q-ml-sm q-mb-none text-grey-8">hotels.search.title</p>
  <hr data-v-b2120978="" class="q-separator q-separator--horizontal bg-grey-6" aria-orientation="horizontal">
  <div data-v-b2120978="" class="q-pa-md full-width column wrap items-end">
    <div data-v-b2120978="" class="inputs">
      <div data-v-b2120978="" outlined="" clearable="" dense="" display-value="" class="full-width"><span class="text-subtitle1 text-grey-6 text-weight-bold">hotels.search.destination <span class="text-red-5">*</span></span><label class="q-field row no-wrap items-start q-field--outlined q-select q-field--auto-height q-select--with-input q-select--without-chips q-select--single q-field--dense full-width" style="width: 100%;" for="f_2086f6f4-b19e-4313-9929-8041c89bfed4">
          <!---->
          <div class="q-field__inner relative-position col self-stretch">
            <div class="q-field__control relative-position row no-wrap" tabindex="-1">
              <div class="q-field__control-container col relative-position row no-wrap q-anchor--skip">
                <div class="q-field__native row items-center"><input class="q-field__input q-placeholder col" type="search" tabindex="0" role="combobox" aria-readonly="false" aria-autocomplete="list" aria-expanded="false" aria-controls="f_2086f6f4-b19e-4313-9929-8041c89bfed4_lb" id="f_2086f6f4-b19e-4313-9929-8041c89bfed4" value=""></div>
                <!---->
              </div>
              <div class="q-field__append q-field__marginal row no-wrap items-center q-anchor--skip"><i class="q-icon notranslate material-icons q-select__dropdown-icon" aria-hidden="true" role="presentation">arrow_drop_down</i></div>
              <!---->
            </div>
            <!---->
          </div>
          <!---->
        </label></div>
      <div data-v-b2120978="" modelvalue="" dense="" class="full-width" outlined="" debounce="400"><span class="text-subtitle1 text-grey-6 text-weight-bold">hotels.search.name <span class="text-red-5">*</span></span><label class="q-field row no-wrap items-start q-field--outlined q-input q-field--dense full-width" for="f_831729be-7ede-4b9d-af41-3d376fceaa2a">
          <!---->
          <div class="q-field__inner relative-position col self-stretch">
            <div class="q-field__control relative-position row no-wrap" tabindex="-1">
              <div class="q-field__control-container col relative-position row no-wrap q-anchor--skip"><input class="q-field__native q-placeholder" tabindex="0" id="f_831729be-7ede-4b9d-af41-3d376fceaa2a" type="text" value="">
                <!---->
              </div>
            </div>
            <!---->
          </div>
          <!---->
        </label></div>
    </div><button data-v-b2120978="" class="q-btn q-btn-item non-selectable no-outline q-btn--standard q-btn--rectangle q-btn--rounded bg-blue text-white q-btn--actionable q-focusable q-hoverable q-btn--no-uppercase q-px-xl q-py-xs" style="font-size: 20px;" tabindex="0" type="button"><span class="q-focus-helper"></span><span class="q-btn__content text-center col items-center q-anchor--skip justify-center row"><span class="block">hotels.search.button</span></span>
      <transition-stub name="q-transition--fade" appear="false" persisted="false" css="true">
        <!---->
      </transition-stub>
    </button>
  </div>
  <transition-stub data-v-b2120978="" enterfromclass="q-transition--fade-enter-from" enteractiveclass="q-transition--fade-enter-active" entertoclass="q-transition--fade-enter-to" leavefromclass="q-transition--fade-leave-from" leaveactiveclass="q-transition--fade-leave-active" leavetoclass="q-transition--fade-leave-to" appear="true" persisted="false" css="true">
    <!---->
  </transition-stub>
</div>"
`;
